mitmproxy/proxy/layers/quic.py

```diff
@@ -372,15 +390,29 @@ def initialize_replacement(peer_cid: bytes) -> None:
 
     quic._initialize = initialize_replacement  # type: ignore
     try:
-        quic.receive_datagram(data, ("0.0.0.0", 0), now=0)
+        for dgm in datagrams:
+            quic.receive_datagram(dgm, ("0.0.0.0", 0), now=time.time())
     except QuicClientHello as hello:
         try:
             return ClientHello(hello.data)
         except EOFError as e:
             raise ValueError("Invalid ClientHello data.") from e
     except QuicConnectionError as e:
         raise ValueError(e.reason_phrase) from e
-    raise ValueError("No ClientHello returned.")
+
+    quic_logger = quic._configuration.quic_logger
+    assert isinstance(quic_logger, QuicLogger)
+    traces = quic_logger.to_dict().get("traces")
+    assert isinstance(traces, list)
+    for trace in traces:
+        quic_events = trace.get("events")
+        for event in quic_events:
+            if event["name"] == "transport:packet_dropped":
+                raise ValueError(
+                    f"Invalid ClientHello packet: {event['data']['trigger']}"
+                )
```

Comment:
Really don't like this part, but I don't think we have any other way around this.

Reply:
`quic_parse_client_hello_from_datagrams` is a strong contender for the worst-function-in-mitmdump award, I'm glad we're strengthening its position!